Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/common/certificate/CertificateProvider.kt

```diff
@@ -2,28 +2,74 @@ package de.rki.coronawarnapp.covidcertificate.common.certificate
 
 import dagger.Reusable
 import de.rki.coronawarnapp.covidcertificate.common.repository.CertificateContainerId
+import de.rki.coronawarnapp.covidcertificate.recovery.core.RecoveryCertificate
 import de.rki.coronawarnapp.covidcertificate.recovery.core.RecoveryCertificateRepository
+import de.rki.coronawarnapp.covidcertificate.recovery.core.RecoveryCertificateWrapper
+import de.rki.coronawarnapp.covidcertificate.test.core.TestCertificate
 import de.rki.coronawarnapp.covidcertificate.test.core.TestCertificateRepository
+import de.rki.coronawarnapp.covidcertificate.test.core.TestCertificateWrapper
+import de.rki.coronawarnapp.covidcertificate.vaccination.core.VaccinatedPerson
+import de.rki.coronawarnapp.covidcertificate.vaccination.core.VaccinationCertificate
 import de.rki.coronawarnapp.covidcertificate.vaccination.core.repository.VaccinationRepository
+import de.rki.coronawarnapp.util.coroutine.AppScope
+import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
+import de.rki.coronawarnapp.util.flow.shareLatest
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.conflate
+import kotlinx.coroutines.flow.distinctUntilChanged
 import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.plus
 import javax.inject.Inject
 
 @Reusable
 class CertificateProvider @Inject constructor(
-    private val vcRepo: VaccinationRepository,
-    private val tcRepo: TestCertificateRepository,
-    private val rcRepo: RecoveryCertificateRepository,
+    vcRepo: VaccinationRepository,
+    tcRepo: TestCertificateRepository,
+    rcRepo: RecoveryCertificateRepository,
+    @AppScope appScope: CoroutineScope,
+    dispatcherProvider: DispatcherProvider
 ) {
 
+    val certificateContainer: Flow<CertificateContainer> = combine(
+        rcRepo.certificates,
+        tcRepo.certificates,
+        vcRepo.vaccinationInfos
+    ) { recoveries, tests, vaccinations -> CertificateContainer(recoveries, tests, vaccinations) }
+        .conflate()
+        .distinctUntilChanged()
+        .shareLatest(scope = appScope + dispatcherProvider.IO)
+
     /**
      * Finds a [CwaCovidCertificate] by [CertificateContainerId]
      * @throws [Exception] if certificate not found
      */
     suspend fun findCertificate(containerId: CertificateContainerId): CwaCovidCertificate {
-        val certificates = rcRepo.cwaCertificates.first() +
-            tcRepo.cwaCertificates.first() +
-            vcRepo.cwaCertificates.first()
-
+        val certificates = certificateContainer.first().allCwaCertificates
         return certificates.find { it.containerId == containerId }!! // Must be a certificate
     }
+
+    data class CertificateContainer(
+        val recoveryCertificates: Set<RecoveryCertificateWrapper>,
+        val testCertificates: Set<TestCertificateWrapper>,
+        val vaccinationInfos: Set<VaccinatedPerson>
```

Comment:
I don't think we need a container for all certificates in the app which use cases are using this ?

Reply:
The use case which I replaced with the container?
Most use-cases only need the cwa certificates, so I could make the rest private. But there are some special cases like the PersonCertificatesProvider, which also needs the VaccinationInfos. So why shouldn't we keep them accessible?